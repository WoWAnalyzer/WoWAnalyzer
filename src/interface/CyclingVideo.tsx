import { useState } from 'react';
import * as React from 'react';

interface Props extends React.VideoHTMLAttributes<HTMLVideoElement> {
  videos: string[];
  randomValue?: number;
}

const CyclingVideo = ({ videos, randomValue = Math.random(), ...otherProps }: Props) => {
  const [currentIndex, setCurrentIndex] = useState(Math.floor(randomValue * videos.length));
  const currentVideo = videos[currentIndex];

  const handleEnded = () => {
    setCurrentIndex((currentIndex + 1) % videos.length);
  };

  return (
    <video
      autoPlay
      muted
      onEnded={handleEnded}
      key={currentVideo} // without this the element doesn't rerender properly and wouldn't start the next video
      {...otherProps}
    >
      <source src={currentVideo} type="video/mp4" />
    </video>
  );
};

export default CyclingVideo;
