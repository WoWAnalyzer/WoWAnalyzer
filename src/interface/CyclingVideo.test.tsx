import renderer from 'react-test-renderer';
import { fireEvent, render, screen } from '@testing-library/react';

import CyclingVideo from './CyclingVideo';

const FAKE_VIDEOS = ['1', '2', '3'];

describe('CyclingVideo', () => {
  beforeAll(() => {
    // see https://github.com/facebook/react/issues/10389
    Object.defineProperty(HTMLMediaElement.prototype, 'muted', {
      set: vi.fn(),
    });
  });

  it('matches snapshot', () => {
    const tree = renderer.create(<CyclingVideo videos={FAKE_VIDEOS} randomValue={0} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('passes unrecognized props', () => {
    const tree = renderer
      .create(<CyclingVideo videos={FAKE_VIDEOS} randomValue={0} style={{ color: 'red' }} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
  describe('randomValue prop', () => {
    it('renders first video when 0 is passed', () => {
      render(<CyclingVideo videos={FAKE_VIDEOS} randomValue={0} data-testid="cycling-video" />);
      const sources = screen.getByTestId('cycling-video').getElementsByTagName('source');
      expect(sources).toHaveLength(1);
      expect(sources[0].src).toEqual('http://localhost:3000/1');
    });

    it('renders second video when 0.5 is passed', () => {
      render(<CyclingVideo videos={FAKE_VIDEOS} randomValue={0.5} data-testid="cycling-video" />);
      const sources = screen.getByTestId('cycling-video').getElementsByTagName('source');
      expect(sources).toHaveLength(1);
      expect(sources[0].src).toEqual('http://localhost:3000/2');
    });

    it('renders third video when 0.999 is passed', () => {
      render(<CyclingVideo videos={FAKE_VIDEOS} randomValue={0.999} data-testid="cycling-video" />);
      const sources = screen.getByTestId('cycling-video').getElementsByTagName('source');
      expect(sources).toHaveLength(1);
      expect(sources[0].src).toEqual('http://localhost:3000/3');
    });
  });
  it('starts the next video when the current ends', () => {
    render(<CyclingVideo videos={FAKE_VIDEOS} randomValue={0} data-testid="cycling-video" />);
    const sourcesBefore = screen.getByTestId('cycling-video').getElementsByTagName('source');
    expect(sourcesBefore).toHaveLength(1);
    expect(sourcesBefore[0].src).toEqual('http://localhost:3000/1'); // sanity
    fireEvent.ended(screen.getByTestId('cycling-video'));
    const sourcesAfter = screen.getByTestId('cycling-video').getElementsByTagName('source');
    expect(sourcesAfter).toHaveLength(1);
    expect(sourcesAfter[0].src).toEqual('http://localhost:3000/2');
  });
  it('restarts when at the end of the list', () => {
    render(<CyclingVideo videos={FAKE_VIDEOS} randomValue={0.999} data-testid="cycling-video" />);
    const sourcesBefore = screen.getByTestId('cycling-video').getElementsByTagName('source');
    expect(sourcesBefore).toHaveLength(1);
    expect(sourcesBefore[0].src).toEqual('http://localhost:3000/3'); // sanity
    fireEvent.ended(screen.getByTestId('cycling-video'));
    const sourcesAfter = screen.getByTestId('cycling-video').getElementsByTagName('source');
    expect(sourcesAfter).toHaveLength(1);
    expect(sourcesAfter[0].src).toEqual('http://localhost:3000/1');
  });
});
