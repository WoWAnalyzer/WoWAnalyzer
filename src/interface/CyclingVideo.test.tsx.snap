// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CyclingVideo matches snapshot 1`] = `
<video
  autoPlay={true}
  muted={true}
  onEnded={[Function]}
>
  <source
    src="1"
    type="video/mp4"
  />
</video>
`;

exports[`CyclingVideo passes unrecognized props 1`] = `
<video
  autoPlay={true}
  muted={true}
  onEnded={[Function]}
  style={
    Object {
      "color": "red",
    }
  }
>
  <source
    src="1"
    type="video/mp4"
  />
</video>
`;
